File: Knapsack.java
// Time Complexity : O(2^n)
// Space Complexity : O(2^n)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :No

// Your code here along with comments explaining your approach
public class Main {
    public static void main(String[] args) {
        int[] weights = new int[] { 10, 20, 30 };
        int[] profit = new int[] { 60, 100, 120 };
        int bagCapacity = 50;
        System.out.println(maxProfit(bagCapacity, weights, profit));
    }

    public static int maxProfit(int bagCapacity, int weights[], int profit[]) {
        int len = weights.length;
        int[][] dp = new int[len + 1][bagCapacity + 1];

        for (int i = 0; i <= bagCapacity; i++) {
            dp[0][i] = 0;
        }

        for (int i = 1; i < dp.length; i++) {
            for (int j = 0; j < dp[0].length; j++) {

                if (j >= weights[i - 1]) {
                    dp[i][j] = Math.max(dp[i - 1][j], (dp[i - 1][j - weights[i - 1]] + profit[i - 1]));
                } else {
                    dp[i][j] = dp[i - 1][j];
                }
            }
        }
        return dp[len][bagCapacity];

    }
}

File: knapsack.go
// Time Complexity : O(n2)
// Space Complexity : O(nc)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
init a 2d array of size arr+1, c+1
first row and col =0
from 1st row prev = lastrow value, curr =0
if capacity is greater that weight of previous then
update curr += profit from remaining weight
update element as max of prev and curr
return last element
*/

package main

import "fmt"

func knap(w []int, p []int, c int) int {
	if c == 0 || len(w) == 0 || len(p) == 0 || len(w) != len(p) {
		return 0
	}
	arr := make([][]int, len(w) + 1)
	for i:= 0;i < len(arr);i++ {
		arr[i] = make([]int, c + 1)
	}

	for i:= 1;i < len(arr);i++ {
		for j:=1;j<= c;j++ {
			prev := arr[i - 1][j]
			curr := 0
			if (c >= w[i - 1 ]) {
				curr = p[i - 1]
				if j - w[i - 1] > 0 {
					curr += arr[i - 1][j - w[i - 1]]
				}
			}
			arr[i][j] = max(prev, curr)
		}
	}

	return arr[len(w)][c]
}

func max(a, b int) int {
	if a>b {
		return a
	}
	return b
}
func MainKnapsack() {
	fmt.Println(knap([]int {1,2,3}, []int{60, 100, 120}, 5))
}
